mqtt/mqtt-bridge/src/pump.rs

```diff
@@ -0,0 +1,326 @@
+
+            debug!(
+                "{} bridge starting egress publication processing for {:?} pump...",
+                bridge_name, pump_type
+            );
+
+            loop {
+                let mut publish_handle = publish_handle.clone();
+                match select(receive_fut.next(), loader_borrow.try_next()).await {
+                    Either::Left((shutdown, _)) => {
+                        debug!("egress pump received shutdown signal");
+                        if let None = shutdown {
+                            error!(message = "unexpected behavior from shutdown signal while signalling bridge pump shutdown")
+                        }
+
+                        debug!("bridge pump stopped");
+                        break;
+                    }
+                    Either::Right((loaded_element, _)) => {
+                        debug!("egress pump extracted publication from store");
+
+                        if let Ok(Some((key, publication))) = loaded_element {
+                            debug!("publishing publication {:?} for egress pump", key);
+                            if let Err(e) = publish_handle.publish(publication).await {
+                                error!(message = "failed publishing publication for bridge pump", err = %e);
+                            }
+
+                            if let Err(e) = persist.remove(key) {
+                                error!(message = "failed removing publication from store", err = %e);
+                            }
+                        }
+                    }
+                }
+            }
+
+            debug!("pumps for {} bridge stopped...", bridge_name);
+        };
+
+        // incoming pump
+        let bridge_name = self.pump_context.bridge_name.clone();
+        let pump_type = self.pump_context.pump_type.clone();
+        let f2 = async move {
+            debug!(
+                "{} bridge starting ingress publication processing for pump {:?}...",
+                bridge_name, pump_type
+            );
+            self.client.handle_events().await;
+        };
+
+        let f1 = f1.fuse();
+        let f2 = f2.fuse();
+        let mut shutdown = shutdown.fuse();
+        pin_mut!(f1);
+        pin_mut!(f2);
+
+        select! {
+            _ = f1 => {
+                error!(message = "publish loop failed and exited for bridge pump");
+            },
+            _ = f2 => {
+                error!(message = "incoming publication loop failed and exited for bridge pump");
+            },
+            _ = shutdown => {
+                if let Err(e) = client_shutdown.shutdown().await {
+                    error!(message = "failed to shutdown incoming publication loop for bridge pump", err = %e);
+                }
+
+                if let Err(e) = loader_shutdown.send(()) {
+                    error!(message = "failed to shutdown publish loop for bridge pump");
+                }
+            },
+        }
```

Comment:
I am getting this clippy error:
```
generally you want to avoid `&mut &mut _` if possible
```

Any idea why?

Reply:
nope, I see this too

Reply:
try to expand macros and see what code it generates. alternatively you can try `fn future::select()`